### PR TITLE
DockerFingerprints API update

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] - Class can be also used for container fingerprints
- [x] - make() methods are accessible via API
- [x] - Extra helper methods, which allow to easily retrieve facets from Docker fingerprints

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java [hunk 1]
```diff
@@ -7,50 +7,158 @@
...
     }
-
-    private static @Nonnull Fingerprint make(@Nonnull Run<?,?> run, @Nonnull String imageId) throws IOException {
-        return Jenkins.getInstance().getFingerprintMap().getOrCreate(run, "<docker-image>", getImageFingerprintHash(imageId));
+    
+    private static @CheckForNull Fingerprint ofNoException(@Nonnull String id) {
+        try {
+            return of(id);
+        } catch (IOException ex) { // The error is not a hazard in CheckForNull logic
+            LOGGER.log(Level.WARNING, "Cannot retrieve a fingerprint for Docker id="+id, ex);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: Should this really be `WARNING`? Seems like it is fine to ask about a nonexistent ID, in which case this should be `FINE`.

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java [hunk 2]
```diff
@@ -7,50 +7,158 @@
...
+    }
+    
+    /**
+     * Get or create a {@link Fingerprint} for the image.
+     * @param run Origin of the fingerprint (if available)
+     * @param id Image ID. Only 64-char full IDs are supported.
+     * @return Fingerprint for the specified ID
+     * @throws IOException Fingerprint load/save error
+     */
+    public static @Nonnull Fingerprint makeForImage(@CheckForNull Run<?,?> run, @Nonnull String id) throws IOException {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: Would recommend just calling this `forImage`, reads better.

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java [hunk 3]
```diff
@@ -7,50 +7,158 @@
...
+        return Jenkins.getInstance().getFingerprintMap().getOrCreate(run, "<docker-container>", getFingerprintHash(id));
     }
 
     /**
+     * Retrieves a facet from the {@link Fingerprint}.
+     * The method suppresses {@link IOException} if a fingerprint loading fails.
+     * @param <TFacet> Facet type to be retrieved
+     * @param id Docker item ID. Only 64-char full IDs are supported
+     * @param facetClass Class to be retrieved
+     * @return First matching facet. Null may be returned if the loading fails
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: Will also be null if there is such a fingerprint but no such facet, right?

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java [hunk 4]
```diff
@@ -7,50 +7,158 @@
...
+     */
+    @SuppressWarnings("unchecked")
+    public static @Nonnull <TFacet extends FingerprintFacet> Collection<TFacet> getFacets
+            (@Nonnull String id, @Nonnull Class<TFacet> facetClass) { 
+        final Fingerprint fp = ofNoException(id);
+        return (fp != null) ? getFacets(fp, facetClass) : CollectionUtils.EMPTY_COLLECTION;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: `Collections.<TFacet>emptySet()`

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java [hunk 5]
```diff
@@ -7,50 +7,158 @@
...
+    }        
+    
+    //TODO: deprecate and use the core's method when it's available
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: So can these be `private` in the meantime?

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: No. <code>IOException</code> may happen when loading an existing fingerprint from the disk. The method returns null for non-existent fingerprints

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java (hunk 3 above)
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: yes. I'll fix javadoc

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java (hunk 5 above)
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: The idea is to provide this method in API till Jenkins core supports it (I have not even created a PR yet)

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: (OK)

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java (hunk 5 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: But it looks like it is only used internally in this class, right? So we could avoid polluting the `docker-commons` API with it?

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: I use this methods in one of plugins under the development, so I would prefer to keep them public

File: src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java [hunk 6]
```diff
@@ -7,50 +7,159 @@
...
+     * @param <TFacet> Facet type to be retrieved
+     * @param id Docker item ID. Only 64-char full IDs are supported
+     * @param facetClass Class to be retrieved
+     * @return First matching facet. Null may be returned if there is no facet
+     *      or if the loading fails
+     */
+    public static @CheckForNull @SuppressWarnings("unchecked")
+            <TFacet extends FingerprintFacet> TFacet getFacet
+            (@Nonnull String id, @Nonnull Class<TFacet> facetClass) { 
+        final Fingerprint fp = ofNoException(id);
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java]: NIT: the funky line wrapping here makes this a tease to read.

**Comment by tfennelly** [comment]
:+1:

**Pull request merged**